### ISSUE TITLE
staking: add a `min_validator_stake` threshold for validator indexing

### ISSUE THREAD

**Issue by plaidfinch** [penumbra-zone/penumbra#2921]
This is blocked on completion of #2909, because it will be simpler to implement once that's done, and also because it will generate horrible merge conflicts if done concurrently.

## The problem

Currently, the staking component loops over all validators, every epoch end. This presents the possibility for an intentional or unintentional denial of service where the amount of I/O and computation per epoch end scales proportionately to the number of validators which have _ever_ been declared, regardless of their state. As a consequence, a malicious actor could declare a large number of validators, paying an up-front cost to do so as a per-definition fee, and permanently slow down the chain, increasing the resource usage required to operate a node every epoch end. Additionally, a long-lived chain will likely see many non-malicious declarations of validators, many of which will not be successful enough to attain a long-lived position in the active set. Because the continuous per-epoch cost to the chain currently scales with the total historical number of validators ever declared, it is currently a mismatch in deterrance for declaring a validator to assess a one-time fee on validator declarations.

## High-level solutions

There are several conceptual ways to fix this:

1. Impose an iterated fee on declared validators, somehow removing them entirely from the chain if this continuous cost is not met.
2. Change the internal implementation of the chain so that computation and I/O does not scale linearly per-epoch with the total historical number of validators.
3. Require a minimum delegation amount for a validator to be considered even to be `Inactive`, and ensure that validators below this minimum amount won't be iterated over. (This is essentially [what Namada does](https://namada.net/blog/namada-cubic-proof-of-stake).)

~~The second option is strongly preferable to the first, because it does not require additional operational concerns for penumbra validators, and it does not complicate the staking state model further by adding yet another state into which a validator can enter.~~ Edit: after further discussion with @hdevalence, we've decided to go with (3), since it addresses the issue while reducing the chance of implementation error now or in the future when modifying the staking implementation. **The revised proposed implementation and analysis is directly below, with the old design and analysis left as a comment to record the alternate proposal.**

As we will see, achieving (2) actually does impose an implicit continuous cost on any validator whose mere validator definition's existence causes a per-epoch cost. (Though this is no longer relevant because we are going with the third option).

## Why was it this way to begin with?

We store the current rate for each validator in the state:

```rust
pub fn current_rate_by_validator(id: &IdentityKey) -> String {
    format!("staking/validator_rate/current/{id}")
}
```

This key stores values of type `RateData`:

```rust
pub struct RateData {
    /// The validator's identity key.
    pub identity_key: IdentityKey,
    /// The index of the epoch for which this rate is valid.
    pub epoch_index: u64,
    /// The validator-specific reward rate.
    pub validator_reward_rate: u64,
    /// The validator-specific exchange rate.
    pub validator_exchange_rate: u64,
}
```

A `RateData` includes the `epoch_index` (above), which means that keeping the current rate... current... requires iterating over every validator, every epoch-end, _even if neither the reward rate and exchange rate changed_ during the epoch, merely to increment the epoch index in the `RateData`.

## Implementation

This implementation is based on a design feature of the Namada staking system, the `min_validator_stake` threshold.

### Avoid tracking the `epoch_index` within `RateData`

These implementation steps set up the possibility of _avoiding updating validators whose rates don't change_.

- [ ] Define a related struct `CurrentRateData` instead of `RateData` to be stored at the key `current_rate_by_validator`, which is an exact copy of `RateData` except it does not contain the `epoch_index` field. This will require defining a new proto for `CurrentRateData` so it can be stored in the state.
- [ ] Define `impl From<RateData> for CurrentRateData`.
- [ ] Move the `RateData::next` method from `RateData` to `CurrentRateData`.
- [ ] Change `set_validator_rates` to take a `CurrentRateData` and store that instead of a `RateData`.
- [ ] Add a new `current_validator_rate` state accessor which returns only the `CurrentRateData`, while renaming the old `current_validator_rate` to `validator_rate` and changing all use sites to use `validator_rate` instead.
- [ ] Modify `validator_rate` so that it _still_ returns a `RateData`, but does so rather by reading both the current epoch index _as well as_ that validator's `CurrentRateData` (gotten using the new `current_rate_data`), and packaging the two together to form a `RateData` incorporating the current epoch index.
- [ ] Inside the main validator reward and slashing calculation loop, use `validator_rate` to get a `RateData`, the `From` impl to downcast it to a `CurrentRateData`, `CurrentRateData::next` to update the rate for each validator, and `set_validator_rates` to write it back. This is slightly circuitous, but it minimizes the amount of other code that needs to be modified to use `CurrentRateData` instead of `RateData`.

#### Correctness notes

We presently update the `epoch_index` field of every `RateData` for every validator to always be the present epoch's index. Now, this is done synthetically instead, preserving the same interface to `validator_rate`, with the same behavior.

### Add a new `Defined` state for validators below a delegation threshold

Now that we can avoid updating validators whose rates don't change, we can define a new initial validator state whose rates definitively won't begin to change until it reaches a certain threshold of delegations.

- [ ] Create a new validator state `Defined` that is the new initial state of a validator upon first upload. Update protos, `DomainType` impls, etc. to accommodate this.
- [ ] Define a new governance parameter `min_validator_stake` which sets a threshold in the staking token of `.unbonded_stake()`-converted delegations which a validator must obtain in order to exit the `Defined` state and enter the `Inactive` or `Disabled` state — and if it ever falls below this threshold, in any state, it will transition back to `Defined`. The details of these state transitions are described in later steps.
- [ ] Modify the action handler for `ValidatorDefinition` actions so that the initial state of a newly defined validator is `Defined` rather than `Disabled` or `Inactive` (this means modifying `add_validator` so that the initial state used is `Defined`, and also following the breadcrumbs inside this function to make sure that the relevant metrics gauges are updated, etc.) Note that this may resolve what looks like a possible bug, where all new validators were set to `Inactive`, even if they were not set to `enabled` — this is done differently in `update_validator`, so this should be fixed with these code changes.
- [ ] Modify `update_validator` so that a `Defined` validator is treated here the same way an `Inactive` one is.
- [ ] Define state transitions for `set_validator_state_inner` as follows:
    - `Defined` => `Inactive`: register the validator in the index of `Inactive` validators
    - `Defined` => `Disabled`: register the validator in the index of `Disabled` validators
    - `Defined` => `Active | Jailed | Disabled | Tombstoned`: impossible because none of these things can happen to a validator without it first passing through `Inactive` or `Disabled`
    - `Inactive` => `Defined`: remove the validator from the index of `Inactive` validators
    - `Disabled` => `Defined`: remove the validator from the index of `Disabled` validators
    - `Active` => `Defined`: impossible because the validator should have first become `Inactive`
    - `Tombstoned` => `Defined`: impossible because `Tombstoned` is the terminal state
- [ ] Add metrics for the `Defined` state and correctly set them in `set_validator_state_inner`.

#### Correctness notes

As discussed in the discord, it is okay if a governance proposal decreases `min_validator_stake` and some validators are still in the `Defined` state when they should be `Inactive` — they will change states to the correct one as soon as any delegation/undelegation "pokes" them. The reverse issue, where a proposal _increases_ the `min_validator_stake` and a validator is still in the `Inactive` state when it should by rights be merely `Defined`, is not serious, but can be averted by a change noted below in control flow and iteration changes which checks for this case during the initial loop over all non-`Defined` validators.

### New indices to maintain

- [ ] For each of the states `Active`, `Inactive`, `Jailed`, `Disabled`, and `Tombstoned`, maintain a separate index of which validator identities are in each of those states by adding code to `set_validator_state` and `add_genesis_validator` (to make sure we track `Active` genesis validators). Do not keep an index of the validators in the `Defined` state, because we should never need to iterate over them. The reason why the indices are separate is to enable future work (see #2909) to iterate efficiently over all `Active` validators, as well as to iterate over `Active` and `Inactive` validators in the processing of active/inactive transitions.
- [ ] Maintain a separate index of validator identities which are currently `Unbonding`, indexed by the unbonding epoch, making sure to preserve lexical ordering so they can be range-enumerated. The way this index should be updated is by modifying it exclusively in `set_validator_bonding_state`, and adding a getter to get a stream of all validators unbonding in a given epoch. This is necessary to efficiently process pending unbondings while including even `Defined` validators.

#### Correctness notes

The only place validator states and bonding states are modified is in their respective setters, so it should be safe to maintain these indices as described.

### Control flow and iteration changes

- [ ] Change the current iteration over _all_ validators in end-epoch to iterate only over all validators which are not in the `Defined` state, by iterating over each of the indices above, in sequence. This should be done by changing the `validator_list` method on the staking component's `StateReadExt` trait to enumerate only validators which are present in the indices, rather than the entire list of validators. While doing this modification, might as well also make it so that this method returns a stream, rather than a `Vec` (and tweak the other places in the staking code where a similar pattern occurs).
- [ ] Also, modify `validator_identity_list` to return only those validators which are not `Defined`, for consistency.
- [ ] Lift out of and _after_ this loop the calculation of delegation changes and voting power: instead of looping over all validators, this should loop the _set union_ of all validators with delegation changes (which we already conveniently have collected in a data structure) and the set of all non-`Defined` validators (so that we can catch the case of a governance parameter increasing the `min_validator_stake` threshold and thereby causing some validators to revert to `Defined` without any delegation changes). In this loop, set state changes `Defined` => `Inactive | Disabled` and `Inactive | Disabled` => `Defined` transitions for validators which are over/under the governance parameter threshold. Which of `Inactive` or `Disabled` a `Defined` validator becomes is dependent on the `enabled` flag in the `Validator` struct associated with its identity key.
- [ ] Pulling things out and ensuring this sequencing also means that the issuance of rewards to funding streams needs to be pulled out into its own loop over all `Active` validators (this can be done efficiently by using the index defined above), performed _after_ the new token supplies for each validator are updated in the state (which happens when processing the delegation deltas, which must happen after the processing of penalties, as described above).
- [ ] Change the implementation of `process_validator_unbondings` to use the index defined for unbondings to only unbond validators scheduled for unbonding in this epoch _or before_ (use a range query to make sure we don't miss ones from before, just in case).
- [ ] Change the implementation of `set_active_and_inactive_validators` to examine only validators which are in the `Active` or `Inactive` states, by using the indices defined for them.

In doing this, there will be numerous opportunities to modularize the code.

#### Correctness notes

**Crucially:** changes to validator state _must_ occur _after_ the main loop over all non-`Defined` validators which assigns slashing penalties, or else a slashed validator's delegators could mass-undelegate before the end of the epoch to bring it below the`min_validator_stake` threshold, thereby escaping the attention of the loop over all validators which assigns slashing penalties.

The changes to implementations suggested which use indices to avoid iterating over more validators do so by changing loops over all validators that filter whether they do anything inside the loop, into loops over only the relevant validators to begin with. The same effects should occur in the refactored versions suggested.

## Performance and spam-deterrance

After we do these changes, there will be 3 newly-modularized loops when processing validator updates in end-epoch:

1. Loop over all non-`Defined` validators, calculating and applying slashing penalties and updating rates and voting powers.
2. Loop over all delegations/undelegations, adjusting the supply of the delegation token and unbonded staking token, and changing the state of validators who cross the `Defined`/`Inactive | Disabled` threshold.
3. Loop over all `Active` validators, calculating and emitting rewards to funding streams.

Then, already modularized into separate functions, there are two more loops, modified by these changes:

4. Loop over only validators `Unbonding` in this particular epoch, as changed above, to set bonding states to `Unbonded` of validators which have unbonded in this epoch.
5. Loop over only `Active` and `Inactive` validators, as changed above, to move the states of validators that transition between those two lists.

The asymptotic bottleneck here has, by these changes, become the first loop over all non-`Defined` validators, and this is prevented economically from growing too large because keeping a validator in this loop requires a delegation of a threshold amount.

**Comment by erwanor** [comment]
Done in #3609, minus the rate data refactor. We should reexamine it after #3615.
